### PR TITLE
Feature/hhausner tpc compression

### PR DESCRIPTION
**Pull request by hausnerh**
### Description

Update ICARUS TPC PhysCrateOverlays to allow for compressed TPC fragments. I add some structures to the overlays which help to keep track of the fragment size and ADC values. By loading in the ADC values one at the creation of the overlay we save on looping that would otherwise be required for non-indexable fragments.

Additionally added methods to read words from artdaq::Fragments safely, where accessing pointers is kept under a separate function to ensure those calls are done safely. These functions are not currently used everywhere, but could be in the future.

### Related Repository Branches

To set the TPC compression in the board readers we need the [hhausner_setTPCCompression branch of sbndaq-artdaq](https://github.com/SBNSoftware/sbndaq-artdaq/pull/135)

Additionally using the overlays to read compressed TPC data requires updates to the ICARUS TPC decoder tools. Those updates are available in the [feature/hhausner_testCompressedTPCFragments_forPullRequest branch of icaruscode](https://github.com/SBNSoftware/icaruscode/tree/feature/hhausner_testCompressedTPCFragments_forPullRequest). Offline access to the updated overlays will be needed for icaruscode to process compressed data.

### Testing details
- *Where it was tested*: SBN-FD
- *Run number associated with test*: 11420
- Other information: These updates preserve the function of the overlays on non-compressed data, and so will not effect ordinary running conditions unless the compression is enabled in the board readers

### CODE REVIEW

**Comment by gennadiy-fnal** [comment]
This pull request accumulates changes that were made during the preceding two years. And hence, reviewing individual code commits is not very practical.

**Comment by hausnerh** [comment]
> This pull request accumulates changes that were made during the preceding two years. And hence, reviewing individual code commits is not very practical.

If you would prefer me to make the necessary changes on a fresh branch I can do so

**Comment by cerati** [comment]
@gennadiy-fnal , should the PR be made against develop?

**Comment by gennadiy-fnal** [comment]
> @gennadiy-fnal , should the PR be made against develop?

@cerati That is an excellent question. Our development model uses the concept of `release branches`, where each release branch is based on the develop branch and is merged into the develop branch once the release is finalized. So, pull requests should be made to the current release branch (release/v1_09_00) and can be merged into it only after we have at least one positive review. Once the code is merged, we schedule the integration test at the detector.

**Comment by gennadiy-fnal** [comment]
> > This pull request accumulates changes that were made during the preceding two years. And hence, reviewing individual code commits is not very practical.
> 
> If you would prefer me to make the necessary changes on a fresh branch I can do so

@hausnerh I'm not certain what the best strategy would be here; perhaps other reviewers can make suggestions. We have other PRs that combine 2 years' worth of work, which makes them impractical to review on a `per-commit` basis. We need to discuss at the next DAQ meeting what the `expectation` is for the review process. I'm not convinced that we can apply the `glance and approve` methodology.

**Comment by cerati** [comment]
since we need these changes urgently for offline, is there a way to speedup the review process (or separate it for the online)?

**Comment by gennadiy-fnal** [comment]
> since we need these changes urgently for offline, is there a way to speedup the review process (or separate it for the online)?

@cerati I'm not directly involved in the offline releases, so you'll need to get in touch with your offline release coordinator.

**Comment by cerati** [comment]
> @cerati I'm not directly involved in the offline releases, so you'll need to get in touch with your offline release coordinator.

@gennadiy-fnal are the online and offline release integration separate? let me add also @miquelnebot

File: for-jenkins/build-sbndaq_artdaq_core.sh
Comment (dtorretta56) [inline for-jenkins/build-sbndaq_artdaq_core.sh]: why the artdaq_version is downgrade here? thanks

File: for-jenkins/publishscisoft.sh
Comment (dtorretta56) [inline for-jenkins/publishscisoft.sh]: also here for lined 19-20, why are the qualifiers 'older' ?

**Review by dtorretta56 — CHANGES_REQUESTED** [review]
before approval, please address my comments. thanks

File: for-jenkins/build-sbndaq_artdaq_core.sh
Comment (hausnerh) [reply for-jenkins/build-sbndaq_artdaq_core.sh]: These come from originally pulling from develop and not v1_09_00. This and the other version skew are fixed.

File: for-jenkins/publishscisoft.sh
Comment (hausnerh) [reply for-jenkins/publishscisoft.sh]: These come from originally pulling from develop and not v1_09_00. This and the other version skew are fixed.

File: for-jenkins/build-sbndaq_artdaq_core.sh
Comment (gennadiy-fnal) [reply for-jenkins/build-sbndaq_artdaq_core.sh]: @dtorretta56  These are the two scripts (build-sbndaq_artdaq_core.sh, publishscisoft.sh) I use for running Jenkins builds, so it's safe to leave them as they are. I'll make adjustments myself if needed.

File: for-jenkins/publishscisoft.sh
Comment (gennadiy-fnal) [reply for-jenkins/publishscisoft.sh]: @dtorretta56 @cerati @hausnerh Those two files are useed by the Jenkins build system, and I often make adjustments to them before creating the release build. They do not impact the Icarus release quality. Therefore, if those are the only two files in question, I recommend that you approve this request.

**Review by cerati — APPROVED** [review]
I tested that this branch works as intended in the offline release, for the purpose of re-compressing raw data files and running the reconstruction on top. I cannot comment on the code change details or the online.

File: for-jenkins/publishscisoft.sh
Comment (dtorretta56) [reply for-jenkins/publishscisoft.sh]: thanks Gennadiy. Then I approve this request.

File: for-jenkins/build-sbndaq_artdaq_core.sh
Comment (dtorretta56) [reply for-jenkins/build-sbndaq_artdaq_core.sh]: Tanks Gennadiy . OK to proceed then.

**Pull request merged**
